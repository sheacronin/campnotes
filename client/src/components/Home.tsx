import React, { useEffect, useState } from 'react';
import { ICampaign } from '../types';
import EditCampaign from './EditCampaign';
import Campaign from './Campaign';
import '../styles/Campaigns.css';

function Home() {
  const [campaigns, setCampaigns] = useState<ICampaign[]>([]);
  const [isAddingCampaign, setIsAddingCampaign] = useState(false);

  useEffect(() => {
    getCampaigns();
  }, []);

  async function getCampaigns() {
    try {
      const response = await fetch('http://localhost:5000/campaigns', {
        credentials: 'include',
      });
      const data = await response.json();
      setCampaigns(data);
    } catch (error) {
      console.error(error);
    }
  }

  return (
    <>
      <h2>Campaigns</h2>
      <section className="campaigns-container">
        {campaigns.map((campaign) => (
          <Campaign
            key={campaign.id}
            campaign={campaign}
            setCampaigns={setCampaigns}
          />
        ))}
        {isAddingCampaign ? (
          <EditCampaign
            setIsEditing={setIsAddingCampaign}
            setCampaigns={setCampaigns}
          />
        ) : (
          <button
            onClick={() => setIsAddingCampaign((prevState) => !prevState)}
          >
            Add Campaign
          </button>
        )}
      </section>
    </>
  );
}

export default Home;
