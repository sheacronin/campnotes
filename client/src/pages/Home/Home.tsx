import React from 'react';
import { ICampaign } from './components/Campaign/useCampaigns';
import EditCampaign from './components/Campaign/EditCampaign';
import Campaign from './components/Campaign/Campaign';
import useEditForm from '../../useEditForm';
import './components/Campaign/Campaigns.css';

interface HomeProps {
  campaigns: ICampaign[];
  addCampaign: (campaign: ICampaign) => void;
  editCampaign: (id: number, campaign: ICampaign) => void;
}

function Home({ campaigns, addCampaign, editCampaign }: HomeProps) {
  const {
    isEditing: isAddingCampaign,
    onSubmitForm,
    openEditForm,
    closeEditForm,
  } = useEditForm(addCampaign, editCampaign);

  return (
    <>
      <h2 className="page-heading">Campaigns</h2>
      <section className="campaigns-container">
        {campaigns.map((campaign) => (
          <Campaign
            key={campaign.id}
            campaign={campaign}
            addCampaign={addCampaign}
            editCampaign={editCampaign}
          />
        ))}
        {isAddingCampaign ? (
          <EditCampaign
            onSubmitForm={onSubmitForm}
            closeEditForm={closeEditForm}
          />
        ) : (
          <button onClick={() => openEditForm()}>Add Campaign</button>
        )}
      </section>
    </>
  );
}

export default Home;
