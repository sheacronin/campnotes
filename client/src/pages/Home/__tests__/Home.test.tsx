import React from 'react';
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import Home from '../Home';

describe('Campaign component', () => {
  const CAMPAIGNS = [
    { id: 1, title: 'Campaign1' },
    { id: 2, title: 'Campaign2' },
    { id: 3, title: 'Campaign3' },
  ];
  const addCampaign = jest.fn();
  const editCampaign = jest.fn();

  it('renders correct page heading', () => {
    render(
      <Home
        campaigns={CAMPAIGNS}
        addCampaign={addCampaign}
        editCampaign={editCampaign}
      />
    );
    expect(
      screen.getByRole('heading', { name: 'Campaigns' }).textContent
    ).toMatch(/Campaigns/);
  });

  it('matches snapshot', () => {
    const { container } = render(
      <Home
        campaigns={CAMPAIGNS}
        addCampaign={addCampaign}
        editCampaign={editCampaign}
      />
    );
    expect(container).toMatchSnapshot();
  });
});
