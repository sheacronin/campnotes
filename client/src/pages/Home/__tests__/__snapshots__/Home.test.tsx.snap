// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Campaign component matches snapshot 1`] = `
<div>
  <h2
    class="page-heading"
  >
    Campaigns
  </h2>
  <section
    class="campaigns-container"
  >
    <article
      class="campaign"
    >
      <h3>
        Campaign1
      </h3>
      <button>
        Edit
      </button>
    </article>
    <article
      class="campaign"
    >
      <h3>
        Campaign2
      </h3>
      <button>
        Edit
      </button>
    </article>
    <article
      class="campaign"
    >
      <h3>
        Campaign3
      </h3>
      <button>
        Edit
      </button>
    </article>
    <button>
      Add Campaign
    </button>
  </section>
</div>
`;
